lib/vtls/schannel.c

```diff
@@ -405,6 +405,362 @@ get_cert_location(TCHAR *path, DWORD *store_name, TCHAR **store_path,
+    schannel_cred.grbitEnabledProtocols = SP_PROT_SSL2_CLIENT;
+    break;
+  default:
+    failf(data, "Unrecognized parameter passed via CURLOPT_SSLVERSION");
+    return CURLE_SSL_CONNECT_ERROR;
+  }
+
+  if(SSL_CONN_CONFIG(cipher_list)) {
+    result = set_ssl_ciphers(&schannel_cred, SSL_CONN_CONFIG(cipher_list));
+    if(CURLE_OK != result) {
+      failf(data, "Unable to set ciphers to passed via SSL_CONN_CONFIG");
+      return result;
+    }
+  }
+
+#ifdef HAS_CLIENT_CERT_PATH
+  /* client certificate */
+  if(data->set.ssl.primary.clientcert) {
+    DWORD cert_store_name;
+    TCHAR *cert_store_path = NULL;
+    TCHAR *cert_thumbprint_str;
+    CRYPT_HASH_BLOB cert_thumbprint;
+    BYTE cert_thumbprint_data[CERT_THUMBPRINT_DATA_LEN];
+    HCERTSTORE cert_store;
+    FILE *fInCert = NULL;
+
+    TCHAR *cert_path = curlx_convert_UTF8_to_tchar(
+                           data->set.ssl.primary.clientcert);
+    if(!cert_path)
+      return CURLE_OUT_OF_MEMORY;
+
+    result = get_cert_location(cert_path, &cert_store_name,
+                               &cert_store_path, &cert_thumbprint_str);
+    if((result != CURLE_OK) && (data->set.ssl.primary.clientcert[0]!='\0'))
+      fInCert = fopen(data->set.ssl.primary.clientcert, "rb");
+
+    if((result != CURLE_OK) && !fInCert) {
+      failf(data, "schannel: Failed to get certificate location"
+            " or file for %s",
+            data->set.ssl.primary.clientcert);
+      curlx_unicodefree(cert_path);
+      return result;
+    }
+
+    if(fInCert) {
+      /* Reading a .P12 or .pfx file, like the example at bottom of
+         https://social.msdn.microsoft.com/Forums/windowsdesktop/
+         en-US/3e7bc95f-b21a-4bcd-bd2c-7f996718cae5
+      */
+      void *certdata = NULL;
+      long filesize = 0;
+      CRYPT_DATA_BLOB datablob;
+      WCHAR* pszPassword;
+      size_t pwd_len = 0;
+      int str_w_len = 0;
+      int continue_reading = fseek(fInCert, 0, SEEK_END) == 0;
+      if(continue_reading)
+        filesize = ftell(fInCert);
+      if(filesize < 0)
+        continue_reading = 0;
+      if(continue_reading)
+        continue_reading = fseek(fInCert, 0, SEEK_SET) == 0;
+      if(continue_reading)
+        certdata = malloc(((size_t)filesize) + 1);
+      if(!certdata ||
+         ((int) fread(certdata, (size_t)filesize, 1, fInCert) != 1))
+        continue_reading = 0;
+      fclose(fInCert);
+      curlx_unicodefree(cert_path);
+
+      if(!continue_reading) {
+        failf(data, "schannel: Failed to read cert file %s",
+              data->set.ssl.primary.clientcert);
+        free(certdata);
+        return CURLE_SSL_CERTPROBLEM;
+      }
+
+      /* Convert key-pair data to the in-memory certificate store */
+      datablob.pbData = (BYTE*)certdata;
+      datablob.cbData = (DWORD)filesize;
+
+      if(data->set.ssl.key_passwd != NULL)
+        pwd_len = strlen(data->set.ssl.key_passwd);
+      pszPassword = (WCHAR*)malloc(sizeof(WCHAR)*(pwd_len + 1));
```

Comment:
Test for `malloc` failure is missing. If `malloc` fails the code will write to address 0 and crash.

Reply:
Done.

Reply:
Also this is existing code moving around.

Reply:
> Also this is existing code moving around.

This doesn't mean we should do it wrong again